### DISCUSSION TITLE
sorting a list by key

### DISCUSSION THREAD

**Discussion by warmuuh** [kcl-lang/kcl#1666] [General]
i try to sort the output of KCL by a given order (similar to konfig repo, but it doesnt seem to be used there):
```
resourceOrder = [
    "ConfigMap",
    "Deployment"
]

lookupOrder = lambda a: any -> int {
    indexOf(resourceOrder, typeof(a))
}

indexOf = lambda list:[any], ele:any -> int {
    found = [i for i, v in list if v == ele]
    found[0] if len(found) > 0 else None
}

res2 = // rendering of kubernetes resources
manifests.yaml_stream(sorted(res2, False, lookupOrder))
```

but apparently, `sorted` doesnt accept a custom key-lookup function although it was mentioned here: https://www.kcl-lang.io/docs/reference/lang/spec/modules#the-built-in-system-package

how can this be done then?

**Comment by warmuuh** [comment]
worked around by implementing mergesort in dsl... for the interested:
```
merge = lambda left:[any], right:[any], keyFn -> [any] {
    right if len(left) == 0 else \
      (left if len(right) == 0 else \
        ([left[0]] + merge(left[1:], right, keyFn) if keyFn(left[0]) < keyFn(right[0]) else [right[0]] + merge(left, right[1:], keyFn)))
}

mergesort = lambda list:[any], keyFn -> [any] {
    mid = int(len(list) / 2)
    list if len(list) <= 1 else \
      merge(mergesort(list[0:mid], keyFn), mergesort(list[mid:], keyFn), keyFn)
}
```

**Comment by Peefy** [comment]
@warmuuh 

Sorry for the late reply, good job! KCL will support additional parameters for sorted functions in the future, just like Python. Besides, If you are willing, you can also contribute this sorting implementation to the community https://github.com/kcl-lang/modules
